docs/architecture-decisions/fractional.md

```diff
@@ -0,0 +1,121 @@
+---
+status: draft
+author: @toddbaert
+created: 2025-06-06
+updated: 2025-06-06
+---
+
+# Fractional Operator
+
+The fractional operator enables deterministic, fractional feature flag distribution.
+
+## Background
+
+Nearly all feature flag systems require pseudorandom assignment support to facilitate key use cases, including experimentation and fractional progressive rollouts.
+Since flagd seeks to implement a full feature flag evaluation engine, such a feature is required.
+
+## Requirements
+
+- **Deterministic**: must be consistent given the same input (so users aren't re-assigned with each page view, for example)
+- **Performant**: must be quick; we want "predictable randomness", but with a relatively low performance cost
+- **Ease of use**: must be easy to use and understand for basic use-cases
+- **Customization**: must support customization, such as specifying a particular context attribute to "bucket" on
+- **Stability**: adding new variants should result in new assignments for as small a section of the audience as possible
+- **Strong avalanche effect**: slight input changes should result in relatively high chance of differential bucket assignment
```

Comment:
I feel that some of the requirements here are the consequences of the hash algorithm we picked, not necessarily requirements.

Wearing the hat of a user, I would like the actual distributions as close as what I specified in the weights.

Reply:
There is an advantage to this variability from an A/B testing perspective. It wasn't a primary reason for choosing this algorithm but it was at least considered when evaluating the various options.